central/productusage/datastore/securedunits/datastore_impl.go

```diff
@@ -28,29 +32,80 @@ type dataStoreImpl struct {
 var _ DataStore = (*dataStoreImpl)(nil)
 
 // Walk returns the object, if it exists from the store.
-func (ds *dataStoreImpl) Walk(ctx context.Context, from *types.Timestamp, to *types.Timestamp, fn func(*storage.SecuredUnits) error) error {
+func (ds *dataStoreImpl) Walk(ctx context.Context, from time.Time, to time.Time, fn func(*storage.SecuredUnits) error) error {
 	if err := sac.VerifyAuthzOK(usageSAC.ReadAllowed(ctx)); err != nil {
 		return errors.Wrap(err, "cannot permit to walk through usage data")
 	}
-	if from == nil {
-		from, _ = types.TimestampProto(time.Time{})
+	if from.IsZero() {
+		from = time.Unix(0, 0)
 	}
-	if to == nil {
-		to = types.TimestampNow()
+	if to.IsZero() {
+		to = time.Now()
 	}
-	if err := ds.store.Walk(ctx, func(record *storage.SecuredUnits) error {
-		if record.GetTimestamp().Compare(from) >= 0 && record.GetTimestamp().Compare(to) < 0 {
-			return fn(record)
+
+	pagination := search.NewPagination().
+		AddSortOption(search.NewSortOption(search.ProductUsageTimestamp)).Limit(page)
+
+	query := search.NewQueryBuilder().AddTimeRangeField(
+		search.ProductUsageTimestamp, from, to).WithPagination(pagination).ProtoQuery()
+
+	for offset := 0; ; offset += page {
```

Comment:
This does one extra query than necessary. We should check that len(units) < page which means that we are on the last page. Only if the exact number of elements matches page then we'll do the extra query which should be rare